encoder.go

```diff
@@ -78,28 +82,65 @@ const (
+// It is highly advised to use the methods on the object to manipulate it and not set the fields manually.
+type WAFObject = bindings.WAFObject
+
+// Encodable represent a type that can encode itself into a WAFObject.
+// The encodable is responsible for using the [pin.Pinner]
+// object passed in the [EncoderConfig] to pin the data referenced by the encoded [bindings.WAFObject].
+// The encoder must also use the [timer.Timer] passed in the [EncoderConfig] to
+// make sure it doesn't spend too much time doing its job.
+// The encoder must also respect the [EncoderConfig] limits and report truncations.
+type Encodable interface {
+	// Encode encodes itself as the WAFObject obj using the provided EncoderConfig and remaining depth allowed.
+	// It returns a map of truncation reasons and their respective actual sizes. If the error returned is not nil
+	// It is greatly advised to return errors from the waferrors package error when it matters.
+	// Outside of encoding the value, it is expected to check for truncations sizes as advised in the EncoderConfig
+	// and to regularly call the EncoderConfig.Timer.Exhausted() method to check if the encoding is still allowed
+	// and return waferrors.ErrTimeout if it is not.
+	// This method is NOT thread-safe.
+	Encode(config EncoderConfig, obj *bindings.WAFObject, depth int) (map[TruncationReason][]int, error)
+}
+
+func newDefaultEncoder(config EncoderConfig) (*encoder, error) {
+	if config.Pinner == nil {
+		return nil, fmt.Errorf("pinner cannot be nil")
+	}
+	if config.Timer == nil {
+		config.Timer, _ = timer.NewTimer(timer.WithUnlimitedBudget())
+	}
+	if config.ContainerMaxSize < 0 {
+		return nil, fmt.Errorf("container max size must be greater than 0")
+	}
+	if config.StringMaxSize < 0 {
+		return nil, fmt.Errorf("string max size must be greater than 0")
+	}
+	if config.ObjectMaxDepth < 0 {
+		return nil, fmt.Errorf("object max depth must be greater than 0")
+	}
+
+	return &encoder{
+		EncoderConfig: config,
+	}, nil
 }
 
-func newLimitedEncoder(pinner pin.Pinner, timer timer.Timer) encoder {
-	return encoder{
-		pinner:           pinner,
-		timer:            timer,
-		containerMaxSize: bindings.MaxContainerSize,
-		stringMaxSize:    bindings.MaxStringLength,
-		objectMaxDepth:   bindings.MaxContainerDepth,
+func newLimitedEncoderConfig(pinner pin.Pinner, timer timer.Timer) EncoderConfig {
+	return EncoderConfig{
+		Pinner:           pinner,
+		Timer:            timer,
+		ContainerMaxSize: bindings.MaxContainerSize,
+		StringMaxSize:    bindings.MaxStringLength,
+		ObjectMaxDepth:   bindings.MaxContainerDepth,
 	}
 }
 
-func newMaxEncoder(pinner pin.Pinner) encoder {
-	timer, _ := timer.NewTimer(timer.WithUnlimitedBudget())
-	return encoder{
-		pinner:           pinner,
-		timer:            timer,
-		containerMaxSize: math.MaxInt,
-		stringMaxSize:    math.MaxInt,
-		objectMaxDepth:   math.MaxInt,
+func newMaxEncoderConfig(pinner pin.Pinner) EncoderConfig {
```

Comment:
This could be `newUnlimitedEncoderConfig` while the other one is just `newEncoderConfig` by contrast?